submission/distributor.go

```diff
@@ -147,22 +161,74 @@ func (d *Distributor) refreshRoots(ctx context.Context) map[string]error {
 	defer d.mu.Unlock()
 
 	d.logRoots = freshRoots
+	d.rootDataFull = len(d.logRoots) == len(d.logClients)
+	// Merge individual root-pools into a unified one
+	d.rootPool = ctfe.NewPEMCertPool()
+	for _, pool := range d.logRoots {
+		for _, c := range pool.RawCertificates() {
+			d.rootPool.AddCert(c)
+		}
+	}
 
 	return errors
 }
 
+// IsRootDataFull returns whether each registered Log has root-info.
+func (d *Distributor) isRootDataFull() bool {
+	d.mu.RLock()
+	defer d.mu.RUnlock()
+	return d.rootDataFull
+}
+
 // SubmitToLog implements Submitter interface.
 func (d *Distributor) SubmitToLog(ctx context.Context, logURL string, chain []ct.ASN1Cert) (*ct.SignedCertificateTimestamp, error) {
-	// TODO(Mercurrent) add implementation.
-	return nil, nil
+	lc, ok := d.logClients[logURL]
+	if !ok {
+		return nil, fmt.Errorf("No client registered for Log with URL %q", logURL)
+	}
+	return lc.AddPreChain(ctx, chain)
 }
 
 // AddPreChain runs add-pre-chain calls across subset of logs according to
 // Distributor's policy. May emit both SCTs array and error when SCTs
 // collected do not satisfy the policy.
 func (d *Distributor) AddPreChain(ctx context.Context, rawChain [][]byte) ([]*AssignedSCT, error) {
-	// TODO(Mercurrent) add implementation.
-	return []*AssignedSCT{}, nil
+	if len(rawChain) == 0 {
+		return nil, fmt.Errorf("Distributor unable to process empty chain")
+	}
+
+	// Validate and root the chain.
+	d.mu.RLock()
+	vOpts := ctfe.NewCertValidationOpts(d.rootPool, time.Time{}, false, false, nil, nil, false, nil)
+	rootedChain, err := ctfe.ValidateChain(rawChain, vOpts)
+
+	if err != nil && d.isRootDataFull() {
+		d.mu.RUnlock()
+		return nil, fmt.Errorf("Distributor unable to process cert-chain: %v", err)
+	}
+	compatibleLogs := d.ll.Compatible(rootedChain, d.logRoots)
+	d.mu.RUnlock()
+
+	// Set-up policy structs.
+	var cert *x509.Certificate
+	if len(rootedChain) != 0 {
+		cert = rootedChain[0]
+	} else {
```

Comment:
My expectations: `ctfe.ValidateChain()` to return an empty chain when validation is failing. No matching roots and/or chain is invalid. MalformedChainRequest-test is covering invalid chain case.

Reply:
Ah ok, so you could alternatively write this as follows and it'd have the same behaviour?
```go
if err == nil {
  cert = rootedChain[0]
} else {
```

Reply:
The above code is a bit tricky to follow. You effectively handle the `err` returned by `ctfe.ValidateChain()` on both lines 205 and 216 (because `len(rootChain) == 0` is synonymous with `err != nil`), while doing something else inbetween (calling `d.ll.Compatible()`). That `Compatible()` call takes `rootedChain` as an argument, which is nil if `err != nil`, but that only results in an error on line 225 if `d.logRoots` is *not* missing roots for a set of logs that meet the policy. It's a tricky set of conditions to keep in mind while reading it. Any ideas for how this could be simplified? If not, just some comments to document these conditions would help I think.

Reply:
I reworked it. Not sure it's simplified enough but it looks a bit more straightforward. WDYT?

Reply:
I'll take a look once you push your changes 🙂

Reply:
LGTM, thanks!